include/glow/Optimizer/GraphOptimizer/NodeSplitting.h

```diff
@@ -0,0 +1,285 @@
+  return [=](const SplitNodeContext &ctx) -> bool {
+    for (const auto *splitNode : ctx.splitNodes) {
+      if (!(splitNode->getTotMemSize() <= maxMem)) {
+        return false;
+      }
+    }
+    return true;
+  };
+}
+
+///===---------------------------------------------------------------------===//
+///                                  splitNode
+///===---------------------------------------------------------------------===//
+/// Function to split the node \p node using the given option \p splitOption
+/// and/or the given split constraint \p splitConstraint. At least the split
+/// option or the split constraint must be given (not necessarily both), the one
+/// which is not used must be set to null. If no option/constraint is provided
+/// then an error is thrown. \returns a vector with the nodes obtained after
+/// splitting \p node. The node is split only if there is a splitting logic
+/// implemented for the node type, otherwise an empty vector is returned.
+Expected<std::vector<Node *>>
+splitNode(Node *node, const SplitNodeOption *splitOption,
+          const SplitNodeConstraint *splitConstraint);
+
+/// Function to split the node \p node using the option \p splitOption.
+/// \returns a vector with the nodes obtained after splitting \p node. The node
+/// is split only if there is a splitting logic implemented for the node type,
+/// otherwise an empty vector is returned.
+Expected<std::vector<Node *>> splitNode(Node *node,
+                                        const SplitNodeOption &splitOption);
+
+/// Function to split the node \p node using the constraint \p splitConstraint.
+/// \returns a vector with the nodes obtained after splitting \p node. The node
+/// is split only if there is a splitting logic implemented for the node type,
+/// otherwise an empty vector is returned.
+Expected<std::vector<Node *>>
+splitNode(Node *node, const SplitNodeConstraint &splitConstraint);
+
+///===---------------------------------------------------------------------===//
+///                                  splitNodes
+///===---------------------------------------------------------------------===//
+/// Split node option map provided to the node splitting procedure which
+/// consists in a mapping between a node and the split option to be used
+/// for that node.
+using SplitNodeOptionMap = llvm::DenseMap<Node *, SplitNodeOption *>;
+
+/// Split node constraint map provided to the node splitting procedure which
+/// consists in a mapping between a node and the split constraint to be used
+/// for that node.
+using SplitNodeConstraintMap = llvm::DenseMap<Node *, SplitNodeConstraint *>;
+
+/// Split node map returned by the node splitting procedure which consists in a
+/// mapping between the original node and the vector of nodes it was split into.
+/// If the node was not actually split then the vector is empty.
+using SplitNodeMap = std::unordered_map<Node *, std::vector<Node *>>;
```

Comment:
I changed the pointers for option and constraint to const to forbid the node splitting functions to change them.
I left the returned map node pointer type to non-const since the user should be allowed to do anything with returned map.